src/java/org/apache/cassandra/locator/DynamicEndpointSnitch.java

```diff
@@ -154,31 +326,203 @@ private void registerMBean()
+        }
 
         MBeanServer mbs = ManagementFactory.getPlatformMBeanServer();
         try
         {
-            mbs.unregisterMBean(new ObjectName(mbeanName));
+            if (mbeanRegistered)
+                mbs.unregisterMBean(new ObjectName(mbeanName));
         }
         catch (Exception e)
         {
             throw new RuntimeException(e);
         }
     }
 
+    /**
+     * Background task running on the samples dictionary. The default implementation sends latency probes (PING)
+     * messages to explore nodes that we have not received timings for recently but have ranked in
+     * {@link DynamicEndpointSnitch#sortedByProximity(InetAddressAndPort, ReplicaCollection)}.
+     */
+    protected void updateSamples()
+    {
+        // Split calculation of probe timers from sending probes for testability
+        calculateProbes(samples, dynamicLatencyProbeInterval);
+
+        if (!StorageService.instance.isGossipActive())
+            return;
+
+        schedulePings(samples);
+    }
+
+    /**
+     * This method mutates the passed AnnotatedMeasurements to implement capped exponential backoff per endpoint.
+     *
+     * The algorithm is as follows:
+     * 1. All samples get their millisSinceLastMeasure and millisSinceLastRequest fields
+     *    incremented by the passed interval
+     * 2. Any recently requested (ranked) endpoints that have not been measured recently (e.g. because the snitch
+     *    has sent them no traffic) get probes with exponential backoff.
+     *
+     * The backoff is capped at MAX_PROBE_INTERVAL_MS. Furthermore the probes are stopped after
+     * MAX_PROBE_INTERVAL_MS of no ranking requests as well.
+     *
+     * At the end of this method, any passed AnnotatedMeasurements that need latency probes will have non zero
+     * probeTimerMillis members set.
+     */
+    @VisibleForTesting
+    static void calculateProbes(Map<InetAddressAndPort, AnnotatedMeasurement> samples, long intervalMillis) {
+        for (Map.Entry<InetAddressAndPort, AnnotatedMeasurement> entry: samples.entrySet())
+        {
+            if (entry.getKey().equals(FBUtilities.getBroadcastAddressAndPort()))
+                continue;
+
+            AnnotatedMeasurement measurement = entry.getValue();
+            long lastMeasure = measurement.millisSinceLastMeasure.getAndAdd(intervalMillis);
+            long lastRequest = measurement.millisSinceLastRequest.getAndAdd(intervalMillis);
+
+            if (lastMeasure >= MIN_PROBE_INTERVAL_MS && lastRequest < MAX_PROBE_INTERVAL_MS)
+            {
+                if (measurement.probeTimerMillis == 0)
+                {
+                    measurement.probeTimerMillis = intervalMillis;
+                }
+                else if (measurement.probeFuture != null && measurement.probeFuture.isDone())
+                {
+                    measurement.probeTimerMillis = Math.min(MAX_PROBE_INTERVAL_MS, measurement.probeTimerMillis * 2);
```

Comment:
So this is going to only schedule the next probe MIN_PROBE_INTERVAL_MS after the previous probe has been sent? So every probe is going to have MIN_PROBE_INTERVAL_MS delay added to it on top of whatever the delay is supposed to be?

Seems like the first probe always goes out after 2x MIN_PROBE_INTERVAL_MS? It takes MIN_PROBE_INTERVAL_MS before we check if it is needed and then it's delayed for 2x MIN_PROBE_INTERVAL_MS when the right thing to do is probe immediately.

Reply:
I take this back. I think being a bounded amount slower at scheduling pings is fine and I like how things are relatively simple right now.

Reply:
Ack, +1